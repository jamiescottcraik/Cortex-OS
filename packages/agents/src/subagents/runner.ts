/**
 * Subagent execution runner with isolation
 *
 * This module provides the execution environment for subagents with proper
 * isolation, context management, and delegation support.
 */

import { v4 as uuidv4 } from 'uuid';
import { CortexAgent } from '../CortexAgent';
import { createLogger } from '../mocks/voltagent-logger';
import { createModelRouter } from '../utils/modelRouter';
import type {
	DelegationRequest,
	DelegationResult,
	ISubagentRunner,
	SubagentContext,
	SubagentResult,
	ToolAccessControl,
} from './types';

const logger = createLogger('SubagentRunner');

/**
 * Tool access control implementation
 */
class SubagentToolAccessControl implements ToolAccessControl {
	constructor(
		private allowed: string[],
		private blocked: string[],
	) {}

	allow: string[] = this.allowed;
	block: string[] = this.blocked;

	isAccessible(toolName: string): boolean {
		// Check block list first
		if (this.blocked.some((pattern) => this.matchPattern(toolName, pattern))) {
			return false;
		}

		// If allow list is empty, allow all (except blocked)
		if (this.allowed.length === 0) {
			return true;
		}

		// Check allow list
		return this.allowed.some((pattern) => this.matchPattern(toolName, pattern));
	}

	private matchPattern(name: string, pattern: string): boolean {
		// Simple glob-style matching
		if (pattern === '*') return true;
		if (pattern.endsWith('*')) {
			return name.startsWith(pattern.slice(0, -1));
		}
		if (pattern.startsWith('*')) {
			return name.endsWith(pattern.slice(1));
		}
		return name === pattern;
	}
}

/**
 * Main subagent runner implementation
 */
export class SubagentRunner implements ISubagentRunner {
	private delegator?: ISubagentDelegator;

	constructor(delegator?: ISubagentDelegator) {
		this.delegator = delegator;
	}

	/**
	 * Execute a subagent with the given context
	 */
	async execute(context: SubagentContext): Promise<SubagentResult> {
		const startTime = Date.now();
		logger.info(`Executing subagent: ${context.config.name} (${context.id})`);

		try {
			// Create tool access control
			const accessControl = new SubagentToolAccessControl(
				context.config.allowed_tools || [],
				context.config.blocked_tools || [],
			);

			// Filter tools based on access control
			const filteredTools = context.tools.filter((tool) =>
				accessControl.isAccessible(tool.name),
			);

			// Create model router with subagent's model preferences
			const modelRouter = createModelRouter({
				enableMLX:
					context.config.model_provider === 'mlx' ||
					!context.config.model_provider,
				ollamaBaseUrl:
					context.config.model_provider === 'ollama'
						? (context.config.model_config?.ollamaBaseUrl as string)
						: undefined,
				apiProviders: this.buildApiProvidersConfig(context.config),
			});

			// Create a temporary CortexAgent instance for this subagent
			const agent = new CortexAgent({
				name: context.config.name,
				description: context.config.description,
				model: context.config.model,
				modelProvider: context.config.model_provider,
				enableMemory: context.config.memory_enabled,
				tools: filteredTools,
				modelRouter,
			});

			// Set up delegation if enabled
			if (this.delegator && context.config.auto_delegate) {
				agent.setDelegator(this.delegator);
			}

			// Execute the agent
			const result = await agent.execute({
				message: context.input,
				context: {
					isolationId: context.config.context_isolation
						? context.id
						: undefined,
					recursionDepth: context.metadata.recursionDepth,
					maxRecursion: context.config.max_recursion,
					scope: context.config.scope,
				},
				options: {
					timeout: context.config.timeout_ms,
					maxTokens: context.config.max_tokens,
				},
			});

			const duration = Date.now() - startTime;

			return {
				success: true,
				output: result.content,
				toolCalls: result.toolCalls,
				metrics: {
					duration,
					tokensUsed: result.tokenUsage?.total,
					toolCalls: result.toolCalls?.length || 0,
				},
			};
		} catch (error) {
			const duration = Date.now() - startTime;
			logger.error(`Subagent execution failed: ${context.config.name}`, error);

			return {
				success: false,
				error: error instanceof Error ? error.message : String(error),
				metrics: {
					duration,
					toolCalls: 0,
				},
			};
		}
	}

	/**
	 * Check if the subagent is healthy
	 */
	async healthCheck(): Promise<boolean> {
		try {
			// Simple health check - could be extended to check model availability
			return true;
		} catch (error) {
			logger.error('Health check failed:', error);
			return false;
		}
	}

	/**
	 * Get subagent capabilities
	 */
	async getCapabilities(): Promise<string[]> {
		return ['text-generation', 'tool-use', 'memory-access', 'delegation'];
	}

	/**
	 * Build API providers configuration from subagent config
	 */
	private buildApiProvidersConfig(config: any) {
		const providers: any = {};

		if (config.model_provider === 'openai' && config.model_config?.apiKey) {
			providers.openai = {
				apiKey: config.model_config.apiKey,
				baseURL: config.model_config.baseURL,
			};
		} else if (
			config.model_provider === 'anthropic' &&
			config.model_config?.apiKey
		) {
			providers.anthropic = {
				apiKey: config.model_config.apiKey,
				baseURL: config.model_config.baseURL,
			};
		} else if (
			config.model_provider === 'google' &&
			config.model_config?.apiKey
		) {
			providers.google = {
				apiKey: config.model_config.apiKey,
				baseURL: config.model_config.baseURL,
			};
		}

		return providers;
	}
}

/**
 * Subagent delegator interface
 */
export interface ISubagentDelegator {
	delegate(request: DelegationRequest): Promise<DelegationResult>;
}

/**
 * Default delegator implementation
 */
export class DefaultSubagentDelegator implements ISubagentDelegator {
	constructor(private runners: Map<string, SubagentRunner>) {}

	async delegate(request: DelegationRequest): Promise<DelegationResult> {
		const startTime = Date.now();
		logger.info(`Delegating to: ${request.to}`);

		const runner = this.runners.get(request.to);
		if (!runner) {
			return {
				success: false,
				error: `Subagent not found: ${request.to}`,
				metrics: {
					duration: Date.now() - startTime,
					agent: request.to,
				},
			};
		}

		// Get the subagent config
		// This is a simplified version - in practice, you'd need to access the registry
		const config = {
			name: request.to,
			description: `Delegated subagent: ${request.to}`,
			scope: 'project' as const,
			auto_delegate: false, // Prevent further delegation
			max_recursion: 0, // No recursion for delegated calls
			timeout_ms: request.metadata?.timeout || 30000,
			context_isolation: true,
			memory_enabled: true,
		};

		const context: SubagentContext = {
			id: uuidv4(),
			config,
			input: request.message,
			tools: [], // Tools would be populated from the registry
			metadata: {
				startTime,
				parentId: 'delegated',
				recursionDepth: 0,
				delegated: true,
			},
		};

		const result = await runner.execute(context);

		return {
			success: result.success,
			response: result.output,
			error: result.error,
			metrics: {
				duration: Date.now() - startTime,
				agent: request.to,
			},
		};
	}
}
