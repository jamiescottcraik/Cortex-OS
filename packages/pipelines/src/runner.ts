import { Worker } from 'node:worker_threads';
import { PipelineManifest } from './manifest.js';
import { isAllowed } from './allowlist.js';

export async function runPipeline(manifest: unknown, input: unknown, timeoutMs = 30000): Promise<unknown> {
  const parsed = PipelineManifest.parse(manifest);
  if (!isAllowed(parsed.id)) {
    throw new Error(`Pipeline ${parsed.id} not allowed`);
  }
  return new Promise((resolve, reject) => {
    const worker = new Worker(parsed.entry, { workerData: input });
    const timer = setTimeout(() => {
      worker.terminate();
      reject(new Error('Pipeline timed out'));
    }, timeoutMs);
    worker.on('message', (msg) => {
      clearTimeout(timer);
      resolve(msg);
    });
    worker.on('error', (err) => {
      clearTimeout(timer);
      reject(err);
    });
  });
}
