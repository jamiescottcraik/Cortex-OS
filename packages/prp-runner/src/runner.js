/**
 * @file packages/prp-runner/src/runner.ts
 * @description End-to-end PRP workflow runner using gate framework (G0â†’G1 for now)
 */
import { generatePRPMarkdown, generateReviewJSON, writePRPDocument, } from './documentation/index.js';
import { loadInitialMd } from './enforcement/initial-processor.js';
import { G0IdeationGate } from './gates/g0-ideation.js';
import { G1ArchitectureGate } from './gates/g1-architecture.js';
import { G2TestPlanGate } from './gates/g2-test-plan.js';
import { G3CodeReviewGate } from './gates/g3-code-review.js';
import { G4VerificationGate } from './gates/g4-verification.js';
import { G5TriageGate } from './gates/g5-triage.js';
import { G6ReleaseReadinessGate } from './gates/g6-release-readiness.js';
import { G7ReleaseGate } from './gates/g7-release.js';
class AutoApproveProvider {
    async requestApproval({ gateId, role, context, }) {
        return {
            gateId: gateId,
            actor: context.actor || role,
            decision: 'approved',
            timestamp: new Date().toISOString(),
            commitSha: context.repoInfo.commitSha,
            rationale: `Auto-approved by ${role} in non-strict mode`,
        };
    }
}
export async function runPRPWorkflow(blueprint, repoInfo, options, approvalProvider) {
    const enforcementProfile = await loadInitialMd(options.projectRoot, options.initialMdPath);
    // Initialize state
    const state = {
        id: `prp-${Date.now()}`,
        runId: `run-${Date.now()}`,
        phase: 'strategy',
        blueprint: {
            title: blueprint.title,
            description: blueprint.description,
            requirements: blueprint.requirements,
            metadata: blueprint.metadata,
        },
        enforcementProfile,
        gates: {},
        approvals: [],
        outputs: {},
        evidence: [],
        validationResults: {},
        metadata: {
            startTime: new Date().toISOString(),
        },
    };
    const gates = [
        new G0IdeationGate(),
        new G1ArchitectureGate(),
        new G2TestPlanGate(),
        new G3CodeReviewGate(),
        new G4VerificationGate(),
        new G5TriageGate(),
        new G6ReleaseReadinessGate(),
        new G7ReleaseGate(),
    ];
    const ctxBase = {
        workingDirectory: options.workingDirectory,
        projectRoot: options.projectRoot,
        enforcementProfile,
        repoInfo,
        actor: options.actor || 'system',
        strictMode: !!options.strictMode,
    };
    const approver = approvalProvider ?? new AutoApproveProvider();
    for (const gate of gates) {
        const context = { ...ctxBase, state };
        const result = await gate.execute(context);
        if (result.requiresHumanApproval) {
            const approval = await approver.requestApproval({
                gateId: gate.id,
                role: gate.humanApprovalSpec?.role || 'reviewer',
                description: gate.humanApprovalSpec?.description || gate.name,
                context,
            });
            state.approvals.push(approval);
            // Attach to gate result in state for completeness
            result.humanApproval = approval;
            // If approved and no automated failures, promote status to passed; if rejected, set to failed
            const hasFailures = result.automatedChecks.some((c) => c.status === 'fail');
            if (approval.decision === 'approved' && !hasFailures) {
                result.status = 'passed';
            }
            else if (approval.decision !== 'approved') {
                result.status = 'failed';
            }
            // If strictMode and approval rejected, stop
            if (options.strictMode && approval.decision !== 'approved') {
                break;
            }
        }
        // Persist final gate result after any approval handling
        state.gates[gate.id] = result;
    }
    // Generate review JSON and prp.md
    const review = generateReviewJSON(state);
    // Calculate final document status
    const allPassed = ['G0', 'G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7'].every((id) => state.gates[id]?.status === 'passed');
    const status = allPassed ? 'ready-for-release' : 'in-progress';
    const prpDoc = {
        id: state.id,
        title: state.blueprint.title,
        repo: `${repoInfo.owner}/${repoInfo.repo}`,
        branch: repoInfo.branch,
        owner: repoInfo.owner,
        created: state.metadata.startTime,
        updated: new Date().toISOString(),
        version: '0.1.0',
        status,
        links: {},
    };
    const markdown = generatePRPMarkdown(state, prpDoc, review);
    const prpPath = options.outputPath || `${options.projectRoot}/prp.md`;
    await writePRPDocument(markdown, prpPath);
    return { state, prpPath, markdown };
}
//# sourceMappingURL=runner.js.map