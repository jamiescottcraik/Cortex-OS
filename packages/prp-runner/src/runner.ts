/**
 * @file packages/prp-runner/src/runner.ts
 * @description End-to-end PRP workflow runner using gate framework (G0â†’G1 for now)
 */

import type { HumanApproval, PRPState } from '@cortex-os/kernel';
import {
	generatePRPMarkdown,
	generateReviewJSON,
	type PRPDocument,
	writePRPDocument,
} from './documentation/index.js';
import { loadInitialMd } from './enforcement/initial-processor.js';
import type { BaseGate, GateContext } from './gates/base.js';
import { G0IdeationGate } from './gates/g0-ideation.js';
import { G1ArchitectureGate } from './gates/g1-architecture.js';
import { G2TestPlanGate } from './gates/g2-test-plan.js';
import { G3CodeReviewGate } from './gates/g3-code-review.js';
import { G4VerificationGate } from './gates/g4-verification.js';
import { G5TriageGate } from './gates/g5-triage.js';
import { G6ReleaseReadinessGate } from './gates/g6-release-readiness.js';
import { G7ReleaseGate } from './gates/g7-release.js';

export interface RepoInfo {
	owner: string;
	repo: string;
	branch: string;
	commitSha: string;
}

export interface RunOptions {
	workingDirectory: string;
	projectRoot: string;
	initialMdPath?: string;
	actor?: string;
	strictMode?: boolean;
	outputPath?: string; // Where to write prp.md
}

export interface Blueprint {
	title: string;
	description: string;
	requirements: string[];
	metadata?: Record<string, unknown>;
}

export interface HumanApprovalProvider {
	requestApproval(input: {
		gateId: string;
		role: string;
		description: string;
		context: GateContext;
	}): Promise<HumanApproval>;
}

class AutoApproveProvider implements HumanApprovalProvider {
	async requestApproval({
		gateId,
		role,
		context,
	}: {
		gateId: string;
		role: string;
		description: string;
		context: GateContext;
	}): Promise<HumanApproval> {
		return {
			gateId: gateId as HumanApproval['gateId'],
			actor: context.actor || role,
			decision: 'approved',
			timestamp: new Date().toISOString(),
			commitSha: context.repoInfo.commitSha,
			rationale: `Auto-approved by ${role} in non-strict mode`,
		};
	}
}

export async function runPRPWorkflow(
	blueprint: Blueprint,
	repoInfo: RepoInfo,
	options: RunOptions,
	approvalProvider?: HumanApprovalProvider,
): Promise<{ state: PRPState; prpPath: string; markdown: string }> {
	const enforcementProfile = await loadInitialMd(options.projectRoot, options.initialMdPath);

	// Initialize state
	const state: PRPState = {
		id: `prp-${Date.now()}`,
		runId: `run-${Date.now()}`,
		phase: 'strategy',
		blueprint: {
			title: blueprint.title,
			description: blueprint.description,
			requirements: blueprint.requirements,
			metadata: blueprint.metadata,
		},
		enforcementProfile,
		gates: {},
		approvals: [],
		exports: {},
		outputs: {},
		evidence: [],
		validationResults: {},
		metadata: {
			startTime: new Date().toISOString(),
		},
	};

	const gates: BaseGate[] = [
		new G0IdeationGate(),
		new G1ArchitectureGate(),
		new G2TestPlanGate(),
		new G3CodeReviewGate(),
		new G4VerificationGate(),
		new G5TriageGate(),
		new G6ReleaseReadinessGate(),
		new G7ReleaseGate(),
	];

	const ctxBase: Omit<GateContext, 'state'> = {
		workingDirectory: options.workingDirectory,
		projectRoot: options.projectRoot,
		enforcementProfile,
		repoInfo,
		actor: options.actor || 'system',
		strictMode: !!options.strictMode,
	};

	const approver = approvalProvider ?? new AutoApproveProvider();

	for (const gate of gates) {
		const context: GateContext = { ...ctxBase, state };
		const result = await gate.execute(context);

		if (result.requiresHumanApproval) {
			const approval = await approver.requestApproval({
				gateId: gate.id,
				role: gate.humanApprovalSpec?.role || 'reviewer',
				description: gate.humanApprovalSpec?.description || gate.name,
				context,
			});
			state.approvals.push(approval);
			// Attach to gate result in state for completeness
			result.humanApproval = approval;
			// If approved and no automated failures, promote status to passed; if rejected, set to failed
			const hasFailures = result.automatedChecks.some(
				(c: { status: 'pass' | 'fail' | 'skip' }) => c.status === 'fail',
			);
			if (approval.decision === 'approved' && !hasFailures) {
				result.status = 'passed';
			} else if (approval.decision !== 'approved') {
				result.status = 'failed';
			}
			// If strictMode and approval rejected, stop
			if (options.strictMode && approval.decision !== 'approved') {
				break;
			}
		}

		// Persist final gate result after any approval handling
		state.gates[gate.id] = result;
	}

	// Generate review JSON and prp.md
	const review = generateReviewJSON(state);
	// Calculate final document status
	const allPassed = ['G0', 'G1', 'G2', 'G3', 'G4', 'G5', 'G6', 'G7'].every(
		(id) => state.gates[id]?.status === 'passed',
	);
	const status = allPassed ? 'ready-for-release' : 'in-progress';
	const prpDoc: PRPDocument = {
		id: state.id,
		title: state.blueprint.title,
		repo: `${repoInfo.owner}/${repoInfo.repo}`,
		branch: repoInfo.branch,
		owner: repoInfo.owner,
		created: state.metadata.startTime,
		updated: new Date().toISOString(),
		version: '0.1.0',
		status,
		links: {},
	};
	const markdown = generatePRPMarkdown(state, prpDoc, review);
	const prpPath = options.outputPath || `${options.projectRoot}/prp.md`;
	await writePRPDocument(markdown, prpPath);

	return { state, prpPath, markdown };
}
