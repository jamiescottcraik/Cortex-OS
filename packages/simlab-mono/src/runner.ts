/**
 * @fileoverview SimLab runner - orchestrates user simulation, agent execution, and judging
 * @version 1.0.0
 * @author Cortex-OS Team
 */

import type { SimScenario, SimResult, SimBatchResult, SimTurn } from './types';
import { UserSimulator } from './user-sim';
import { AgentAdapter, PRPExecutor, AgentResponse } from './agent-adapter';
import { Judge } from './judge';
import { SimReporter } from './report';

export interface SimRunnerConfig {
  deterministic?: boolean;
  seed?: number;
  maxTurns?: number;
  timeout?: number;
  debug?: boolean;
  executor?: PRPExecutor;
}

export class SimRunner {
  private readonly userSim: UserSimulator;
  private readonly agentAdapter: AgentAdapter;
  private readonly judge: Judge;
  private readonly reporter: SimReporter;
  private readonly config: SimRunnerConfig;
  private readonly rng: () => number;
  private runCounter: number;

  constructor(config: SimRunnerConfig = {}) {
    this.config = {
      deterministic: true,
      maxTurns: 10,
      timeout: 30000,
      debug: false,
      ...config,
    };

    this.rng =
      this.config.deterministic && typeof this.config.seed === 'number'
        ? this.createSeededRNG(this.config.seed)
        : Math.random;

    this.userSim = new UserSimulator(this.config);
    this.agentAdapter = new AgentAdapter(this.config.executor);
    this.judge = new Judge();
    this.reporter = new SimReporter();
    this.runCounter = this.config.seed ?? 0;
  }

  /**
   * Run a single simulation scenario
   */
  async runScenario(scenario: SimScenario): Promise<SimResult> {
    const runId = this.generateRunId(scenario.id);
    const startTime = Date.now();

    try {
      // Initialize conversation with user simulation
      const initialMessage = await this.userSim.generateInitialMessage(scenario);

      const turns: SimTurn[] = [
        {
          role: 'user' as const,
          content: initialMessage,
          timestamp: new Date().toISOString(),
        },
      ];

      // Conversation loop
      let turnCount = 0;
      let lastResponse = initialMessage;

      while (turnCount < (this.config.maxTurns || 10)) {
        // Agent responds via PRP
        const agentResponse = await this.agentAdapter.execute({
          scenario,
          conversationHistory: turns,
          userMessage: lastResponse,
        });

        turns.push({
          role: 'agent' as const,
          content: agentResponse.content,
          timestamp: new Date().toISOString(),
        });

        // Check if conversation should end
        if (this.shouldEndConversation(agentResponse, scenario)) {
          break;
        }

        // User simulator responds
        const userResponse = await this.userSim.generateResponse(
          scenario,
          turns,
          agentResponse.content,
        );

        if (!userResponse) {
          break; // User simulation indicates conversation end
        }

        turns.push({
          role: 'user' as const,
          content: userResponse,
          timestamp: new Date().toISOString(),
        });

        lastResponse = userResponse;
        turnCount++;
      }

      // Judge the conversation
      const judged = await this.judge.evaluate(scenario, turns);
      const finalResult: SimResult = {
        scenarioId: judged.scenarioId,
        runId,
        passed: judged.passed,
        scores: judged.scores,
        judgeNotes: judged.judgeNotes,
        failures: judged.failures,
        turns: judged.turns,
        metadata: {
          duration: Date.now() - startTime,
          seed: this.config.seed,
          version: '1.0.0',
        },
        timestamp: new Date().toISOString(),
      };
      return finalResult;
    } catch (error) {
      return {
        scenarioId: scenario.id,
        runId,
        passed: false,
        scores: { goal: 0, sop: 0, brand: 0, factual: 0 },
        judgeNotes: `Simulation failed: ${error instanceof Error ? error.message : 'Unknown error'}`,
        failures: ['simulation_error'],
        turns: [],
        timestamp: new Date().toISOString(),
      };
    }
  }

  /**
   * Run multiple scenarios as a batch
   */
  async runBatch(scenarios: SimScenario[]): Promise<SimBatchResult> {
    const batchId = this.generateBatchId();
    const results: SimResult[] = [];

    for (const scenario of scenarios) {
      const result = await this.runScenario(scenario);
      results.push(result);
    }

    return this.reporter.createBatchResult(batchId, results);
  }

  private shouldEndConversation(agentResponse: AgentResponse, scenario: SimScenario): boolean {
    // Check if agent indicated completion
    if (agentResponse.completed) {
      return true;
    }

    // Check if success criteria are met based on agent response
    return (scenario.success_criteria || []).some((criteria: string) =>
      agentResponse.content.toLowerCase().includes(criteria.toLowerCase()),
    );
  }

  private generateRunId(scenarioId: string): string {
    const idPart = this.config.deterministic ? this.runCounter++ : Date.now();
    const suffix = this.config.deterministic ? 'det' : 'rnd';
    return `${scenarioId}-${idPart}-${suffix}`;
  }

  private generateBatchId(): string {
    const rand = Math.floor(this.rng() * 1e9).toString(36);
    return `batch-${rand}`;
  }

  private createSeededRNG(seed: number): () => number {
    let state = seed;
    return () => {
      state = (state * 9301 + 49297) % 233280;
      return state / 233280;
    };
  }
}
