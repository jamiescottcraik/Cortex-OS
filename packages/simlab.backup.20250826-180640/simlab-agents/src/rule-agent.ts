import type { LocalEnv } from "@cortex-os/simlab-env/local-counter";
export type RuleAgent = { decide(s: unknown): Promise<"inc"|"dec"> };
export function greedyToTarget(): RuleAgent {
  return { async decide(s) { const { n, target } = s as LocalEnv["reset"] extends ()=>Promise<infer X> ? X : never; return n < target ? "inc" : "dec"; } };
}

