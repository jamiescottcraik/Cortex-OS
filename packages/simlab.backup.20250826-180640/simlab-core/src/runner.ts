import type { Scenario, Transition } from "@cortex-os/simlab-contracts/scenario";
import { mulberry32 } from "./rng.js";

export type EnvLike = { reset(): Promise<any>; step(a:any): Promise<{ ctx:any; reward:number; done:boolean }> };
export type AgentLike = { decide(s:any): Promise<any> };
export type RunResult = { scenarioId: string; transitions: Transition[]; totalReward: number; seed: number };

export async function runScenario(s: Scenario, env: EnvLike, agent: AgentLike): Promise<RunResult> {
  const rnd = mulberry32(s.seed.value); rnd(); // burn one for parity
  const transitions: Transition[] = [] as any;
  let state = await env.reset(); let total = 0;
  for (let t = 0; t < s.steps; t++) {
    const action = await agent.decide(state);
    const { ctx, reward, done } = await env.step(action);
    transitions.push({ t, state, action, reward, done } as any);
    total += reward; state = ctx; if (done) break;
  }
  return { scenarioId: s.id, transitions, totalReward: Number(total.toFixed(6)), seed: s.seed.value };
}

