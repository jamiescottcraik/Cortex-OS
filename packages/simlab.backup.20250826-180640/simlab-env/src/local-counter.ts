export type Ctx = { n: number; target: number };
export type LocalEnv = {
  reset(): Promise<Ctx>;
  step(a: "inc" | "dec"): Promise<{ ctx: Ctx; reward: number; done: boolean }>;
};
export function counterEnv(init = { start: 0, target: 3 }): LocalEnv {
  let ctx: Ctx = { n: init.start, target: init.target };
  return {
    async reset() { ctx = { n: init.start, target: init.target }; return ctx; },
    async step(a) {
      ctx.n = a === "inc" ? ctx.n + 1 : ctx.n - 1;
      const done = ctx.n === ctx.target;
      const reward = done ? 1 : -0.01;
      return { ctx, reward, done };
    }
  };
}

