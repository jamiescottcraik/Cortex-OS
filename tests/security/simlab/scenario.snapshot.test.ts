import { greedyToTarget } from "@cortex-os/simlab-agents/rule-agent";
import { runScenario } from "@cortex-os/simlab-core/runner";
import { counterEnv } from "@cortex-os/simlab-env/local-counter";
import { describe, expect, it } from "vitest";

describe("scenario snapshots", () => {
	it("matches snapshot for deterministic run", async () => {
		const scenario = {
			id: "snap",
			steps: 5,
			seed: { value: 2024 },
			agent: { id: "a", kind: "rule" },
			env: { id: "e", kind: "local-counter" },
		} as any;
		const res = await runScenario(
			scenario,
			counterEnv({ start: 0, target: 2 }),
			greedyToTarget(),
		);
		expect(res).toMatchSnapshot();
	});
});
