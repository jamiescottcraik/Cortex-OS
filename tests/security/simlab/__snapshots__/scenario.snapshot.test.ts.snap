// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`scenario snapshots matches snapshot for deterministic run 1`] = `
{
  "scenarioId": "snap",
  "transitions": [
    {
      "t": 0,
      "state": {
        "n": 0,
        "target": 2
      },
      "action": "inc",
      "reward": -0.01,
      "done": false
    },
    {
      "t": 1,
      "state": {
        "n": 1,
        "target": 2
      },
      "action": "inc",
      "reward": 1,
      "done": true
    }
  ],
  "totalReward": 0.99,
  "seed": 2024
}
`;
